### PR TITLE
Feat/notification scheduling

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
This PR should demonstrate correct handling of the job scheduling, rescheduling and cancellation, without actually sending any emails.

Dependent on https://github.com/Health-Education-England/TIS-OPS/pull/716

TIS21-5154
TIS21-5155
TIS21-5156

### CODE REVIEW

**Comment by ReubenRobertsHEE** [comment]
Here is an example log output for a PM starting on 4 Dec 2023. Note that week 1 and week 0 notifications are scheduled as normal, while week 4 (being the most recent missed notification) is scheduled 'promptly' (in this case, 1 minute in the future, to demonstrate that it is then picked-up and acted on).

```
2023-11-15T10:37:23.495Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.e.ProgrammeMembershipListener  : Handling programme membership update event ProgrammeMembershipEvent[tisId=7ab1aae3-83c2-4bb6-b1f3-99146e79b362, recrd=RecordDto(data={tisId=7ab1aae3-83c2-4bb6-b1f3-99146e79b362, personId=47165, programmeTisId=1110, programmeName=ACCS Anaesthetics/Peninsula, programmeNumber=SWP905, managingDeanery=Health Education England South West, programmeMembershipType=MILITARY, startDate=2023-12-04, endDate=2028-07-01, programmeCompletionDate=2027-07-04, curricula=[{"curriculumTisId":"40","curriculumName":"ACF (NIHR funding)","curriculumSubType":"ACFNIHR_FUNDING","curriculumSpecialty":"Academic","curriculumMembershipId":"414003","curriculumStartDate":"2023-08-04","curriculumEndDate":"2026-08-04"},{"curriculumTisId":"33","curriculumName":"ACCS Anaesthetics","curriculumSubType":"MEDICAL_CURRICULUM","curriculumSpecialty":"Acute Care Common Stem - Anaesthetics","curriculumMembershipId":"403870","curriculumStartDate":"2023-08-04","curriculumEndDate":"2027-07-04"}], conditionsOfJoining=null}, metadata={})].
2023-11-15T10:37:23.507Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Removed stale notification scheduled for PROGRAMME_UPDATED_WEEK_8-7ab1aae3-83c2-4bb6-b1f3-99146e79b362
2023-11-15T10:37:23.515Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Removed stale notification scheduled for PROGRAMME_UPDATED_WEEK_4-7ab1aae3-83c2-4bb6-b1f3-99146e79b362
2023-11-15T10:37:23.523Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Removed stale notification scheduled for PROGRAMME_UPDATED_WEEK_1-7ab1aae3-83c2-4bb6-b1f3-99146e79b362
2023-11-15T10:37:23.530Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Removed stale notification scheduled for PROGRAMME_UPDATED_WEEK_0-7ab1aae3-83c2-4bb6-b1f3-99146e79b362
2023-11-15T10:37:23.530Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Programme membership 7ab1aae3-83c2-4bb6-b1f3-99146e79b362: excluded false.
2023-11-15T10:37:23.572Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Notification for PROGRAMME_UPDATED_WEEK_4-7ab1aae3-83c2-4bb6-b1f3-99146e79b362 scheduled for Wed Nov 15 10:38:23 GMT 2023
2023-11-15T10:37:23.604Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Notification for PROGRAMME_UPDATED_WEEK_1-7ab1aae3-83c2-4bb6-b1f3-99146e79b362 scheduled for Mon Nov 27 00:00:00 GMT 2023
2023-11-15T10:37:23.636Z  INFO 35200 --- [ntContainer#3-2] u.n.t.t.n.s.ProgrammeMembershipService   : Notification for PROGRAMME_UPDATED_WEEK_0-7ab1aae3-83c2-4bb6-b1f3-99146e79b362 scheduled for Mon Dec 04 00:00:00 GMT 2023
2023-11-15T10:38:23.561Z  INFO 35200 --- [eduler_Worker-2] u.n.t.t.n.service.NotificationService    : Sent DEFAULT.PROGRAMME_UPDATED_WEEK_4-7ab1aae3-83c2-4bb6-b1f3-99146e79b362 notification for 7ab1aae3-83c2-4bb6-b1f3-99146e79b362
```

**Comment by ReubenRobertsHEE** [comment]
The saved History document for the 4-week sent notification

image

**Comment by ReubenRobertsHEE** [comment]
Still draft since the DB _id probably needs fixing, but there may be other suggestions for rework that would be useful to get now.

File: src/main/resources/templates/email/programme-updated-week-0/v1.0.0.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<!DOCTYPE html>
+<html lang="en" xmlns:th="http://www.thymeleaf.org">
+<head>
+  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+</head>
+<body>
+<h1>Email Message</h1>
+<h2>Subject</h2>
+<th:block th:fragment="subject">Programme checklist: 8 weeks to go</th:block>
```
Comment (ReubenRobertsHEE) [inline src/main/resources/templates/email/programme-updated-week-0/v1.0.0.html:9]: These (0-, 1-, and 4- week notification templates) are just placeholders at this point (all copies of the 8-week notification)

File: src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java [hunk 2]
```diff
@@ -69,4 +100,233 @@ public boolean isExcluded(ProgrammeMembership programmeMembership) {
...
+   * @return The notification scheduling date and time.
+   */
+  private Date getScheduleDate(LocalDate programmeStartDate, int daysBeforeStart) {
+    Date milestone;
+    LocalDate milestoneDate = programmeStartDate.minusDays(daysBeforeStart);
+    if (!milestoneDate.isAfter(LocalDate.now())) {
+      // 'Missed' milestones: schedule to be sent soon, but not immediately
+      // in case of human editing 'jitter'.
+      milestone = Date.from(Instant.now()
+          .plus(PAST_MILESTONE_SCHEDULE_DELAY_HOURS, ChronoUnit.HOURS));
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java:249]: The reason for this is to cover scenarios like the following: the LO creates (or edits) a PM and sets the start date to 6 weeks in the future by mistake. They then correct the date to 12 weeks (for example). If notifications were sent immediately, the trainee would receive a 8-week notification, and then no follow--up correction. By delaying the initial notification by an hour, it allows the LO to make the correction, which would then unschedule the incorrect immediate 8-week notification and set it (hopefully correctly) for 4 weeks in the future

File: src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java [hunk 3]
```diff
@@ -69,4 +100,233 @@ public boolean isExcluded(ProgrammeMembership programmeMembership) {
...
+   * Helper function to determine whether a notification should be scheduled.
+   *
+   * @return true if it should be scheduled, false otherwise.
+   */
+  private boolean shouldScheduleNotification(LocalDate programmeStartDate,
+      NotificationType milestone, Map<NotificationType, Instant> notificationsAlreadySent) {
+
+    //do not resend any notification
+    if (notificationsAlreadySent.containsKey(milestone)) {
+      return false;
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java:271]: A/C #3 :
Given that we have already sent reminder to PGDiT
When the start date is changed to be outside 8 weeks 
Then we don't send the reminder email when we get back to the 8 weeks

File: src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java [hunk 4]
```diff
@@ -69,4 +100,233 @@ public boolean isExcluded(ProgrammeMembership programmeMembership) {
...
+      // A future notification, to be scheduled at the appropriate time
+      // unless a more recent notification has already been sent
+      return (notificationsAlreadySent.keySet().stream()
+          .noneMatch(t -> getNotificationDaysBeforeStart(t) < daysBeforeStart));
+    } else {
+      // A past ('missed') notification, to be scheduled promptly
+      // if it is the most recent missed notification.
+      // Do not schedule if it is not the most recent missed notification, regardless of whether
+      // any more recent ones have been sent or not.
+      return (getPastNotifications(programmeStartDate).stream()
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java:287]: A/C 4:
Given that PGDiT has placement starting more then 8 weeks away
When the start date is changed to be within 8 & 4 weeks 
Then send the reminder email

File: src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java [hunk 5]
```diff
@@ -69,4 +100,233 @@ public boolean isExcluded(ProgrammeMembership programmeMembership) {
...
+    }
+
+    Integer daysBeforeStart = getNotificationDaysBeforeStart(milestone);
+    LocalDate milestoneDate = programmeStartDate.minusDays(daysBeforeStart);
+
+    if (milestoneDate.isAfter(LocalDate.now())) {
+      // A future notification, to be scheduled at the appropriate time
+      // unless a more recent notification has already been sent
+      return (notificationsAlreadySent.keySet().stream()
+          .noneMatch(t -> getNotificationDaysBeforeStart(t) < daysBeforeStart));
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java:281]: A/C 2:
Given that PGDiT has placement starting more then 8 weeks away
When the start date is changed within the 4 weeks of the original start date 
Then the reminder email is not sent

**Comment by ReubenRobertsHEE** [comment]
The other A/C's relate to whether to send notifications based on whether they've signed up to TSS, signed their CoJ, etc. which as far as I know we agreed to hold-over to the next slice.

**Comment by DorisWongNHS** [comment]
LGTM

**Review by DorisWongNHS — APPROVED** [review]

**Pull request merged**